[DSS] DDP-7931: User registration endpoint

Trimax opened the pull request:
This PR contains a couple of things:
1. Implementation of the new endpoint for creation the user
2. Some code cleanup (unused methods and constants were removed)
3. UserProfile class builder class was replaced with Lombok's implementation

bskinner committed:
Switch to parameterized logs instead of concat

bskinner committed:
Reflectively call platform's heapDump method

This fixes a bug where compilation could fail due to missing platform management symbols on OpenJ9-descended JVM implementations.

Commit:
TEMPORARY

Commit:
UserProfile -> lombok builder

Commit:
User profile creation

Commit:
Redundant removed

Commit:
DB inserts

Trimax commented on pepper-apis/dss-core/src/main/java/org/broadinstitute/ddp/route/UserCreationRoute.java:
```diff
@@ -0,0 +1,65 @@
+            }
+
+            final var clientId = handle.attach(ClientDao.class).getClientDao().insertClient(
+                    "DON'T KNOW", "DON'T KNOW", tenantDto.getId(), "DON'T KNOW");
```
How exactly should I create the user without knowing all this auth0 stuff? Should it be sent together with payload from the client?

bskinner replied:
We shouldn't be doing anything with the Auth0 client here. Setting up Auth0 clients is done during study initialization and, regardless, any users created via this API will not be bound to an Auth0 account.

Trimax commented on pepper-apis/dss-core/src/main/java/org/broadinstitute/ddp/route/UserCreationRoute.java:
```diff
@@ -0,0 +1,65 @@
+            final var clientId = handle.attach(ClientDao.class).getClientDao().insertClient(
+                    "DON'T KNOW", "DON'T KNOW", tenantDto.getId(), "DON'T KNOW");
+
+            final var user = handle.attach(UserDao.class).createUser(clientId, "DON'T KNOW");
```
How exactly should I create the user without knowing all this auth0 stuff? Should it be sent together with payload from the client?

bskinner replied:
Users created for a Center do not have an associated Auth0 account.

Trimax commented on pepper-apis/dss-database/src/main/resources/db-changes/schema/DDP-7931.xml:
```diff
@@ -0,0 +1,42 @@
+            <column name="study_id" type="bigint">
+                <constraints references="umbrella_study(umbrella_study_id)" foreignKeyName="center_study_study_fk"/>
+            </column>
+        </createTable>
```
This table is required in case the creation flow will be different. In case we will use some operator id instead of center id.

MocanaAtBroad replied:
Maybe this should be "study_center" instead of "center_study"?

bskinner commented on pepper-apis/dss-core/src/main/java/org/broadinstitute/ddp/route/CreateTemporaryUserRoute.java:
```diff
@@ -77,9 +77,10 @@ public Object handle(Request request, Response response, CreateTemporaryUserPayl
-                        .setPreferredLangId(languageId)
+                handle.attach(UserProfileDao.class).createProfile(UserProfile.builder()
+                        .userId(user.getId())
+                        .preferredLangId(languageId)
+                        .preferredLangCode(null)
```
We should be setting one of `preferredLangId` or `preferredLangCode`, but not both (even though it is `null`)- is there a way we can avoid this sort of duplicated call?

Trimax replied:
I was thinking about it. And since we're not using custom builder anymore, I guess the answer is no. But on the other side, if we compare the huge builder class that can be replaced with the one annotation and some places with two calls, I would vote for the 2nd option.

bskinner replied:
That would be fine if these two value were not related. Being able to set both allows the profile to be put into an invalid state.

bskinner commented on pepper-apis/dss-database/src/main/resources/changelog-master.xml:
```diff
@@ -2,7 +2,7 @@
                    xmlns:xsi="http://www.w3.org/2001/XMLSchema-instance"
                    xsi:schemaLocation="http://www.liquibase.org/xml/ns/dbchangelog
-                   http://www.liquibase.org/xml/ns/dbchangelog/dbchangelog-3.5.xsd">
+                   https://www.liquibase.org/xml/ns/dbchangelog/dbchangelog-3.5.xsd">
```
Why is this scheme being changed to `https`, but none of the others are?

Trimax replied:
Looks like I just accepted IntelliJ's suggestion

bskinner replied:
Did some digging into this and the other values are technically opaque identifiers, and not URLs- so this is correct (thanks, XML).

bskinner commented on pepper-apis/dss-core/src/main/java/org/broadinstitute/ddp/DataDonationPlatform.java:
```diff
@@ -368,6 +367,7 @@ private static void start() throws MalformedURLException {
         }
 
         post(API.REGISTRATION, new UserRegistrationRoute(interpreter, new TaskPubSubPublisher()), responseSerializer);
+        post(API.USER_SPECIFIC, new UserCreationRoute(), responseSerializer);
```
If your intention is to add a new post endpoint at `/user`, I don't think this is going to do what you want it do to. The `USER_SPECIFIC` includes a `:userGuid` path parameter and, if this is creating a user, we do not have a GUID yet.

bskinner commented on pepper-apis/dss-core/src/main/java/org/broadinstitute/ddp/json/UserCreationPayload.java:
```diff
@@ -0,0 +1,33 @@
+    @SerializedName("sex")
+    private UserProfile.SexType sex;
+
+    @SerializedName("centerId")
+    private Long centerId;
```
Why is `centerId` a `Long` here? Typically, all external identifiers (such as user guids, activity codes, instance identifiers, etc) are all opaque strings.

bskinner replied:
Spoke with @SimoneMaiwald- this will be transitioned to an opaque string guid when Centers are implemented (like what we're doing in other objects).

bskinner committed:
userdao mods

bskinner committed:
Update User model

* Transition to Lombok @Data
* Auth0UserId and Auth0TenantId are optional now

bskinner committed:
Handle optional Auth0 UserId

bskinner committed:
Adds an email field to the user profile

bskinner committed:
further cleanup of the user class

bskinner committed:
Makes Auth0 tenant and user id fields optional

bskinner committed:
Revert the UserProfile email changes

In the case of internal accounts, the email is a value on-part with the Auth0 user ID, so move it into the user object

bskinner committed:
Handle the User email in the database

bskinner committed:
Rough outline of the Creation response

bskinner committed:
Adds email field to the User

bskinner committed:
Creation route tweaks

bskinner committed:
Adds a new error code for insufficient authz

bskinner committed:
Code cleanup on the user creation updates

bskinner committed:
Adds a User update method for non-required fields

bskinner committed:
Raise some noise if the email isn't there

bskinner committed:
Allows in-memory profiles to be created.

bskinner committed:
Comment cleanup

bskinner committed:
Adds the commons-validator for email address checks

bskinner committed:
Adds a couple of services

Used for managing participant creation & study registration

bskinner committed:
Wrap everything up

bskinner committed:
Removes an unused import

bskinner committed:
Use DDP's pre-configured builder for request deserialzation

bskinner committed:
Minor tweaks

bskinner committed:
**** Disabling authz for testing

bskinner committed:
Exposes the token's issuer

The `iss` claim is the same as the `DDP_TENANT_CLAIM` claim in practice.

bskinner committed:
Adds registration notifications

bskinner committed:
Bug fixes

bskinner committed:
Code cleanup

bskinner committed:
*** TESTING CHANGES

bskinner committed:
DDP-7931: User registration endpoint

bskinner committed:
fixup: merge with email changes

bskinner committed:
Allow users to be created without a created-by clientId

* Fixes a typo in the dss-database main changelog

bskinner commented on pepper-apis/dss-core/src/main/java/org/broadinstitute/ddp/db/dao/CenterProfileDao.java:
```diff
@@ -0,0 +1,17 @@
+
+import java.util.Optional;
+
+public interface CenterProfileDao extends SqlObject {
+    @UseStringTemplateSqlLocator
+    @SqlQuery("SELECT * FROM `center_profile` WHERE center_id = :id")
+    @RegisterConstructorMapper(CenterProfileDto.class)
+    Optional<CenterProfileDto> findById(@Bind("id") final Long id);
+}
```
The Center-related code is out-of-scope for this ticket since none of the management or study-related configuration code is in place. Leave this for future expansion, but it's not currently used.

bskinner commented on pepper-apis/dss-core/src/main/java/org/broadinstitute/ddp/util/JavaHeapDumper.java:
```diff
@@ -1,60 +1,161 @@
+import lombok.extern.slf4j.Slf4j;
+
 @Slf4j
 public class JavaHeapDumper {
```
This class was modified in order to fix compilation issues when using non-HotSpot JVMs (this was preventing my builds). This method isn't called during normal operation, and only called manually when debugging server memory issues so this shouldn't have a significant impact on the overall operation of the backend.

bskinner commented on pepper-apis/dss-core/src/main/java/org/broadinstitute/ddp/route/UserCreationRoute.java:
```diff
@@ -0,0 +1,168 @@
+    public Object handle(Request request, Response response, UserCreationPayload payload) throws Exception {
+        var auth = RouteUtil.getDDPAuth(request);
+
+        if (false && auth.isAuthenticated() == false) {
```
Known `if` bypass here- currently working on testing out the authentication issues. Expect this line to be updated once I get I reliable source of DSM tokens.

MocanaAtBroad replied:
I would say we should trust inside of the route that authentication info is A-OK. This work happening in filters/separate concern.

bskinner replied:
The plan is to remove a lot of this code once I figure out how we're processing the DSM tokens. From what @ssettipalli is telling me, we aren't creating any DDPAuth objects so this code may need to fall-back to manually decoding the token.

bskinner replied:
@MocanaAtBroad Code removed

bskinner commented on pepper-apis/dss-core/src/main/java/org/broadinstitute/ddp/route/UserCreationRoute.java:
```diff
@@ -0,0 +1,168 @@
+        } else {
+            /* LIE */
+            auth = new DDPAuth("https://domain.example", "abcTESTCLIENTdef", "OPERATOR", "totallyAValidJWT", null, "en");
+        }
```
Another block of code that is going away once testing can be performed using DSM tokens.

bskinner commented on pepper-apis/dss-core/src/main/java/org/broadinstitute/ddp/route/UserCreationRoute.java:
```diff
@@ -0,0 +1,168 @@
+        // configured properly.
+        //
+        // Could this be replaced with an "isDSM" check?
+        if (false && auth.hasAdminAccessToStudy(studyGuid) == false) {
```
Another block of code that is going away once testing can be performed using DSM tokens.

bskinner commented on pepper-apis/dss-core/src/main/java/org/broadinstitute/ddp/route/UserCreationRoute.java:
```diff
@@ -0,0 +1,168 @@
+                        String.format("Auth0 client '%s' is not authorized for '%s'.",
+                            requestorClientId, domain));
+                throw ResponseUtil.haltError(HttpStatus.SC_UNAUTHORIZED, error);
+                */
```
Needs to be re-enabled after testing can be performed with real tokens.

Trimax commented on pepper-apis/dss-core/src/main/java/org/broadinstitute/ddp/db/dao/UserProfileSql.java:
```diff
@@ -19,7 +19,7 @@ public interface UserProfileSql extends SqlObject {
-            + "        preferred_language_id, time_zone, do_not_contact, is_deceased, skip_language_popup)"
+            + "        preferred_language_id, time_zone, do_not_contact, is_deceased, skip_language_popup, email)"
             + " values (:userId, :firstName, :lastName, :sex, :birthDate,"
             + " :langId, :tz, :doNotContact, :isDeceased, :skipLanguagePopup)")
```
An e-mail was added to the columns list, but the value seems to be missing

Trimax replied:
In the list of arguments as well

bskinner replied:
Nice catch! Thank you!

Trimax commented on pepper-apis/dss-core/src/main/java/org/broadinstitute/ddp/db/dto/UserDto.java:
```diff
@@ -37,6 +41,14 @@ public class UserDto {
+
+    public Optional<String> getEmail() {
+        return Optional.ofNullable(email);
+    }
```
Do you think it is useful? I see that this method is being used only in one place - UserCreationRoute class.

The construction
```
model.getEmail().orElse(null);
```
looks a bit less readable for me that 
```
Optional.ofNullable(model.getEmail()).orElse(null);
```
or simple ternary operator
```
model.getEmail() == null ? null : model.getEmail();
```

Anyway, it's just a matter of taste, so if you believe returning optional is better, then let's leave it as is.

bskinner replied:
The reasoning for the `Optional<>` here was several-fold:
  * Make it explicit that this was a value that may not be present, and force callers to deal with it (instead of using implicit null handling)
  * Identify places where the method was called, in conjunction with the IDE's "Find References" functionality

While I'd prefer to have this be an Optional, I do agree that this is awkward to use given the current style of our codebase (`model.getEmail().orElse(null);` is also an anti-pattern which should be avoided).

After this latest milestone is reached- let's have a discussion in the process meeting, and figure out what style guidelines we'd like to use in relation to Optionals.

Trimax commented on pepper-apis/dss-core/src/main/java/org/broadinstitute/ddp/json/UserCreationResponse.java:
```diff
@@ -0,0 +1,40 @@
+    public UserCreationResponse(User user, UserProfile profile) {
+        this.guid = user.getGuid();
+        this.hruid = user.getHruid();
+        this.email = user.getEmail().orElseThrow(() -> new IllegalStateException("participant must have a valid email"));
```
We're saying **valid** here, but we don't check the e-mail to be valid. Just not null. Isn't it a bit misleading?

bskinner replied:
Nope. Valid in this case is "non-null". I'd be open to changing the text of the exception to make it clearer, but it is not this object's responsibility to validate a User's email. If a User was created with an invalid email, that's a bug elsewhere.

Trimax commented on pepper-apis/dss-core/src/main/java/org/broadinstitute/ddp/route/AdminLookupInvitationRoute.java:
```diff
@@ -63,12 +63,16 @@ public Object handle(Request request, Response response, LookupInvitationPayload
                 String loginEmail = null;
-                if (user.getAuth0UserId() != null) {
+                
+                if (user.getAuth0UserId().isPresent()) {
```
Regarding returning optionals.
Not sure, but if we'll define the method `hasAuth0UserId` the code will look like
```
if (user.hasAuth0UserId()) {
  ... = user.getAuth0UserId(); //without additional .get statement.
}
```

bskinner replied:
Didn't notice that method! Did a bit of refactoring.

Trimax commented on pepper-apis/dss-core/src/main/java/org/broadinstitute/ddp/route/UserCreationRoute.java:
```diff
@@ -0,0 +1,168 @@
+    public Object handle(Request request, Response response, UserCreationPayload payload) throws Exception {
+        var auth = RouteUtil.getDDPAuth(request);
+
+        if (false && auth.isAuthenticated() == false) {
```
:)
Looks like the work here isn't finished yet.

bskinner replied:
Yup- see [this comment](https://github.com/broadinstitute/ddp-study-server/pull/1653/files#r902675285)

Trimax commented on pepper-apis/dss-core/src/main/java/org/broadinstitute/ddp/route/UserCreationRoute.java:
```diff
@@ -0,0 +1,168 @@
+                var error = new ApiError(ErrorCodes.NOT_FOUND,
+                        String.format("Auth0 client '%s' is not authorized for '%s'.",
+                            requestorClientId, domain));
+                throw ResponseUtil.haltError(HttpStatus.SC_UNAUTHORIZED, error);
```
Should this be uncommented?

bskinner replied:
See [this comment](https://github.com/broadinstitute/ddp-study-server/pull/1653/files#r902677146)

Trimax commented on pepper-apis/dss-core/src/main/java/org/broadinstitute/ddp/service/UserDeleteService.java:
```diff
@@ -368,12 +368,16 @@ private void deleteElasticSearchData(Handle handle, User user, UserCollectedData
-            if (result.hasFailure()) {
-                throw new DDPException(result.hasThrown() ? result.getThrown() : result.getError());
-            }
+        if (user.getAuth0UserId().isPresent() == false) {
```
Let's better use isEmpty instead?
`if (user.getAuth0UserId().isEmpty())`

bskinner replied:
Yup! That's far more straightforward- fixed.

Trimax commented on pepper-apis/dss-core/src/main/java/org/broadinstitute/ddp/service/participants/ParticipantsCreateService.java:
```diff
@@ -0,0 +1,182 @@
+                    TaskPubSubPublisher.TASK_PARTICIPANT_REGISTERED,
+                    payload, studyGuid, participantGuid);
+        */
+    }
```
I don't know as well. Shouldn't it be passed as the constructor argument?

bskinner replied:
Either that or delegated out- either way I don't immediately see where the object is being created (hence my comment), and I don't know what it's lifecycle looks like.

Trimax commented on pepper-apis/dss-core/src/main/java/org/broadinstitute/ddp/service/studies/StudiesService.java:
```diff
@@ -0,0 +1,125 @@
+
+    private boolean studyExists(String studyGuid) {
+        var studyId = handle.attach(JdbiUmbrellaStudy.class).getIdByGuid(studyGuid);
+        if (studyId.isPresent()) {
```
Maybe just
`return studyId.isPresent()`

Or even
`return handle.attach(JdbiUmbrellaStudy.class).getIdByGuid(studyGuid).isPresent();`

bskinner replied:
I'd go with `return studyId.isPresent();` (mostly because it's easier to step through with a debugger). This whole block is the result of a logical compartmentalization between the Optional check, and the method's returned value.

Trimax commented on pepper-apis/dss-core/src/main/java/org/broadinstitute/ddp/util/JavaHeapDumper.java:
```diff
@@ -1,60 +1,161 @@
 package org.broadinstitute.ddp.util;
```
Is this class expected to be in this PR?
Or do you like my way of concomitant unrelated changes? :)

bskinner replied:
See [this comment](https://github.com/broadinstitute/ddp-study-server/pull/1653/files#r902673172)

bskinner committed:
Initial API updates

MocanaAtBroad commented on pepper-apis/dss-core/src/main/java/org/broadinstitute/ddp/DataDonationPlatform.java:
```diff
@@ -334,9 +335,7 @@ private static void start() throws MalformedURLException {
-        before("*", (Request request, Response response) -> {
-            MDC.put(MDC_STUDY, RouteUtil.parseStudyGuid(request.pathInfo()));
-        });
+        before("*", (Request request, Response response) -> MDC.put(MDC_STUDY, RouteUtil.parseStudyGuid(request.pathInfo())));
```
Not saying this is not a good idea. It is! But dropped in the middle of a 72 file PR that has a completely different purpose without a comment is not the best way to make this change clear and transparent.

bskinner replied:
Agreed- this was a change in the PR prior to my work.

Trimax replied:
It wasn't dropped, it was refactored. I just removed braces

bskinner committed:
Code review changes

MocanaAtBroad reviewed (commented):
Still reviewing, but want to submit the comment requesting we remove authentication checks from here. Authorization? I am less sure of that/hope can also be handled in filters.

bskinner committed:
Email was moved to the User table

bskinner committed:
Review comments

bskinner commented:
Everyone, please leave the "Update branch" button alone.

MocanaAtBroad reviewed (commented):
Still reviewing, but want to submit the comment requesting we remove authentication checks from here. Authorization? I am less sure of that/hope can also be handled in filters.

MocanaAtBroad commented on pepper-apis/dss-core/src/main/java/org/broadinstitute/ddp/route/UserCreationRoute.java:
```diff
@@ -0,0 +1,168 @@
+        */ 
+        final var requestorClientId = auth.getClient();
+        final var operatorId = auth.getOperator();
+        final var domain = auth.getIssuer();
```
I see this property is being picked up from token. Do we know if we get this from DSM tokens @ssettipalli ?

MocanaAtBroad commented on pepper-apis/dss-core/src/main/java/org/broadinstitute/ddp/route/UserCreationRoute.java:
```diff
@@ -0,0 +1,168 @@
+        }
+
+        final var email = payload.getEmail();
+        final var emailValidator = EmailValidator.getInstance();
```
I believe that if you just use the @Email annotation on the payload POJO, email will be checked for syntax and automatically rejected by ValidatedJsonInputRoute, so we probably don't need to use EmailValidator explicitly.

MocanaAtBroad commented on pepper-apis/dss-core/src/main/java/org/broadinstitute/ddp/route/UserCreationRoute.java:
```diff
@@ -0,0 +1,168 @@
+            throw ResponseUtil.haltError(HttpStatus.SC_UNPROCESSABLE_ENTITY, error);
+        }
+
+        if (StringUtils.isNotBlank(payload.getCenterId())) {
```
Same sort of validation here. You probably can use @NotEmpty (Part of the purpose of ValidatedJsonInputRoute is to check validation annotations).

https://javadoc.io/doc/javax.validation/validation-api/latest/index.html

bskinner replied:
This isn't validation  of the input- `centerId` being `null` is perfectly ok, we just don't handle the the case where it is not null yet.

MocanaAtBroad commented on pepper-apis/dss-core/src/main/java/org/broadinstitute/ddp/service/participants/ParticipantsCreateService.java:
```diff
@@ -0,0 +1,182 @@
+             * checking the database. If/when Centers get an separate external
+             * identifier (such as a GUID), then this will need to be updated.
+             */
+            Long tempCenterGuid = Long.valueOf(centerGuid);
```
Let's not do this (will not matter for demo as we are not using centerId) but let's use codes/guids whatherver for centers/avoid

MocanaAtBroad commented on pepper-apis/dss-core/src/main/java/org/broadinstitute/ddp/service/participants/ParticipantsCreateService.java:
```diff
@@ -0,0 +1,182 @@
+        var studyDto = handle.attach(JdbiUmbrellaStudy.class).findByStudyGuid(studyGuid);
+        var userDao = handle.attach(JdbiUser.class);
+        var operator = userDao.findByUserGuid(operatorGuid);
+        var participant = userDao.findByUserGuid(participantGuid);
```
We can use the  of JdbiUmbrellaStudyCached here.

MocanaAtBroad commented on pepper-apis/dss-core/src/main/java/org/broadinstitute/ddp/service/studies/StudiesService.java:
```diff
@@ -0,0 +1,120 @@
+    }
+
+    private boolean studyExists(String studyGuid) {
+        var studyId = handle.attach(JdbiUmbrellaStudy.class).getIdByGuid(studyGuid);
```
JdbiUmbrellaStudyCached. Look in code for how it is instantiated (handle.attach is not how you instantiate Cached Jdbi classes)

MocanaAtBroad reviewed (commented):
Not seeing test code for the UserCreationRoute.

bskinner committed:
Fixes a typo

bskinner committed:
This import should not be there

bskinner committed:
Checkstyle fixes

bskinner committed:
Include the email in the SELECT

bskinner committed:
Review comments & further cleanup from auth0 changes

bskinner committed:
Fix another test

bskinner committed:
First pass at fixing a Mockito error

null may not be the correct return value here, but the return value isn't used anyway.

bskinner committed:
Fixes a test

bskinner committed:
SQL updates to get the User's email

bskinner committed:
Code cleanup & comment addition

bskinner committed:
get the study admin filter in place

bskinner committed:
Merge branch 'develop' into DDP-7931
